[user triton] JIT inductor support for new host-side TMA api

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155896
* #155879
* #155827
* __->__ #155814
* #155777

This PR adds JIT inductor support for user-defined triton kernels using the new host-side TMA api.

* handle TensorDescriptor.from_tensor in ir.py
* codegen TensorDescriptor.from_tensor in wrapper.py
* generate the right signature for functions that take TensorDescriptor arguments (i.e. in the @triton_heuristics.user_autotune decorator)

AOTI support is not implemented yet.

Tests: ran test_triton_kernels.py w/ both Triton 3.3 and 3.4 and there were no failures.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

davidberard98 committed:
[user triton] JIT inductor support for new host-side TMA api

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] JIT inductor support for new host-side TMA api"


This PR adds JIT inductor support for user-defined triton kernels using the new host-side TMA api.

* handle TensorDescriptor.from_tensor in ir.py
* codegen TensorDescriptor.from_tensor in wrapper.py
* generate the right signature for functions that take TensorDescriptor arguments (i.e. in the triton_heuristics.user_autotune decorator)

AOTI support is not implemented yet.

Tests: [TODO run tests & get results]

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/common.py:
```diff
@@ -282,6 +282,9 @@ class ConstexprArg:
 @dataclasses.dataclass
 class TMADescriptorArg:
     name: str
+    api_type: str  # "experimental" or "stable"
```
Wondering: are these indeed called "experimental" and "stable" in the latest Triton code?

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1367,7 +1367,7 @@ def _generate_extern_kernel_out_helper(
             self.writeline(f"{kernel}({', '.join(args)})")
 
-    def _generate_tma_descriptor_call(self, desc, apply_size_hints=False):
+    def _generate_tma_descriptor_call_experimental(self, desc, apply_size_hints=False):
```
Can we move this (and the above) to methods of the respective classes?

davidberard98 replied:
does "the respective classes" mean the `ir.TMADescriptorStable` and `ir.TMADescriptorExperimental`? I believe the issue with this is that different wrapper code needs to handle this differently (e.g. this PR implements wrapper code for Python; later PRs implement cpp wrapper code)

aakhundov commented on torch/_inductor/ir.py:
```diff
@@ -6123,26 +6157,35 @@ def __init__(
+            constant_args=block_shape,
+        )
+
+    def get_tensor(self) -> IRNode:
+        return self.tensor
```
The implementation is the same. Can we move this to the parent class?

davidberard98 replied:
good catch, I'll do that

aakhundov reviewed (approved):
Thanks for adding this!

davidberard98 committed:
Update on "[user triton] JIT inductor support for new host-side TMA api"


This PR adds JIT inductor support for user-defined triton kernels using the new host-side TMA api.

* handle TensorDescriptor.from_tensor in ir.py
* codegen TensorDescriptor.from_tensor in wrapper.py
* generate the right signature for functions that take TensorDescriptor arguments (i.e. in the triton_heuristics.user_autotune decorator)

AOTI support is not implemented yet.

Tests: ran test_triton_kernels.py w/ both Triton 3.3 and 3.4 and there were no failures.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] JIT inductor support for new host-side TMA api"


This PR adds JIT inductor support for user-defined triton kernels using the new host-side TMA api.

* handle TensorDescriptor.from_tensor in ir.py
* codegen TensorDescriptor.from_tensor in wrapper.py
* generate the right signature for functions that take TensorDescriptor arguments (i.e. in the triton_heuristics.user_autotune decorator)

AOTI support is not implemented yet.

Tests: ran test_triton_kernels.py w/ both Triton 3.3 and 3.4 and there were no failures.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
